kep/9-capacity-scheduling-for-ml-dl-workloads-based-on-scheduler-framework/README.md

```diff
@@ -0,0 +1,208 @@
+no other Pods can be created - even if the Pod failed to be scheduled, or failed on running (error
+when pulling image, etc.).
+
+2. Improve overall resource utilization by allowing the workloads from one quota to "borrow" unused reserved "min" resources from other quotas. A quota's unused "min" resource can be used by other users, under the condition that there is a mechanism to guarantee the "victim" user can consume his "min" resource whenever he needs. Typically, this is done by implementing preemption.
+
+The success of “ElasticQuota” based capacity scheduling relies on a proper Pod preemption algorithm implementation. This KEP proposes the minimal scheduler extension to support the “ElasticQuota” based scheduling based on the scheduler framework.
+
+
+### Goals
+
+- The API definetion of ElasticQuota
+- Implement a scheduler plugin to honor the API of elastic resource quota.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+A "CRD" is needed to interact with the scheduler. The CRD kind name is subject to change. We use "ElasticQuota" in the proposal. The CRD defines the min and max allocation for the managed resources such as cpu, memory or extended resources. "ElasticQuota" is namespace scoped. A sample yaml is listed below:
+
+```yaml
+apiVersion: v1
+kind: ElasticQuota
+metadata:
+  name: test
+  namespace: test
+spec:
+  max:
+    cpu: 20
+    memory: 40Gi
+    nvidia.com/gpu: 2
+  min:
+    cpu: 10
+    memory: 20Gi
+    nvidia.com/gpu: 1
+```
+
+The definitions of "min" and "max" have been explained above. Max is infinite by default, min is 0 by default. The sum of min in all ElasticQuota must be less than or equal to the total resource capacity of the cluster.The ElasticQuota objects are created by a cluster administrator. One ElasticQuota object per namespace.
+
+### Relationship with ResourceQuota
+This version of proposal does not discuss merging the ElasticQuota into the existing ResourceQuota. But in the future, we will try to merge ElasticQuota into ResourceQuota when it's mature. We have three fields of resource in ElasticQuota(min max) and ResourceQuota(hard). 
+- hard: it will check the resource in pod creation through an admission check. It will prevent submitting too many jobs at one time. The usage of hard is aggregated based on the resource configurations of all pods so it will be greater than or equal to max. 
+- max: it will check the resource in the Prefilter of Pod Scheduling Cycle. The usage of max is aggregated based on the resource configurations of successfully scheduled pods.
+- min: it means the guaranteed resource of Quota.
+
+They act on different phases of pod lifecycle. So it maybe possible to merge them together.
+
+
+### User Stories (optional)
```

Comment:
@denkensk  (cc @Huang-Wei) In our use cases, we want to restrict where a namespace/user can "steal" resources from. A typical scenario is that excessive resources can only be dynamically shared within a team, i.e., among the namespaces belonging to the same team. To support this requirement, conceptually we'll need a two-level queue or a virtual namespace group and each queue/namespace belongs to a high level queue or group. Resource stealing/custom preemption will be only within a group or a top level queue. This would be an important feature in large clusters with several thousands nodes and multiple-tenants, in which t's not practical to randomly steal resources and perform preemption. The latest heretical namespace proposal can be a long term solution.  Thoughts?

Reply:
I understand the concern, but at this moment, due to the lack of official definition of "tenant", it's hard to associate it the elastic RQ with a virtual concept. So IMO we can start with namespace, and try other multi-tenancy APIs such as hierarchical namespace, etc, at a later time.

Reply:
> we can start with namespace, and try other multi-tenancy APIs such as hierarchical namespace, etc, at a later time.
---
I agree with this. In the first version, we can divide the ElasticQuota based on the namespace. We can try to define a higher level of quota to associate with multi-tenant like heretical namespace or virtual cluster from Guo Fei at a later time.

Reply:
However, would it be possible to support a two-level structure, which will cover most use cases. For example, within each namespace, there can be multiple queues with min and max quotas.  I hope it won't introduce much complexity.